luwak/src/main/java/uk/co/flax/luwak/BatchMonitor.java

```diff
@@ -0,0 +1,333 @@
+     *
+     * All queries that take longer than t nanoseconds to run will be recorded in
+     * the slow log.  The default is 2,000,000 (2 milliseconds)
+     *
+     * @param limit the limit in nanoseconds
+     *
+     * @see Matches#getSlowLog()
+     */
+    public void setSlowLogLimit(long limit) {
+        monitor.setSlowLogLimit(limit);
+    }
+
+    @Override
+    public void close() throws IOException {
+        monitor.close();
+    }
+
+    private void scheduleCommit() {
+        // If we are not bunching up commits, then this is a no-op.
+        if (commitTimeout == 0) {
+            return;
+        }
+
+        // If we have a pending commit job, then we are done,
+        // that pending commit will pick up the changes
+        if (commitJob != null && !commitJob.isDone()) {
+            return;
+        }
+
+        // It maybe that lastCommitTime + commitTimeout - System.currentTimeMillis()
+        // is negative, which is fine, the job will just run ASAP then.
+        if (commitTimeout > 0) {
+            commitJob = this.commitExecutor.schedule(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        {
+                            String[] deletes = pendingDeletes.toArray(new String[0]);
+                            if (pendingDeletes.removeAll(Arrays.asList(deletes))) {
+                                monitor.deleteById(deletes);
+                            }
+                        }
+                        {
+                            String[] keys = pendingUpdates.keySet().toArray(new String[0]);
+                            Indexable[] updates = pendingUpdates.values().toArray(new Indexable[0]);
+                            if (pendingUpdates.keySet().removeAll(Arrays.asList(keys))) {
+                                monitor.commit(Arrays.asList(updates));
+                            }
+                        }
+                        lastCommitTime = System.currentTimeMillis();
+                        commitJob = null;
+                    }
+                    catch (Exception e) {
+                        // TODO: How to deal with exceptions here?
+                    }
+                }
+            }, lastCommitTime + commitTimeout - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+        }
+    }
+
+    /**
+     * Add new queries to the monitor
+     * @param queries the MonitorQueries to add
+     * @return a list of exceptions for queries that could not be added
+     * @throws IOException on IO errors
+     */
+    public List<QueryError> update(Iterable<MonitorQuery> queries) throws IOException {
+        if (commitTimeout == 0) {
+            return monitor.update(queries);
+        } else {
+            List<QueryError> errors = new ArrayList<>();
+
+            for (MonitorQuery query : queries) {
+                try {
+                    for (QueryCacheEntry queryCacheEntry : monitor.decomposeQuery(query)) {
```

Comment:
I mean that you can just store the whole query in pendingUpdates, rather than running decomposition here.  It feels to me like this is unnecessary exposure of an internal Monitor process.